swarm/src/lib.rs

```diff
@@ -283,26 +293,69 @@ where TBehaviour: NetworkBehaviour<ProtocolsHandler = THandler>,
         me.network.dial(&addr, handler.into_node_handler_builder()).map(|_id| ())
     }
 
-    /// Tries to reach the given peer using the elements in the topology.
+    /// Tries to initiate a dialing attempt to the given peer, if its status
+    /// is satisfied by the given condition.
+    ///
+    /// If there is an ongoing dialing attempt, the current addresses of the
+    /// peer, as reported by [`NetworkBehaviour::addresses_of_peer`] are added
+    /// to the ongoing dialing attempt, ignoring duplicates.
     ///
-    /// Has no effect if we are already connected to that peer, or if no address is known for the
-    /// peer.
-    pub fn dial(me: &mut Self, peer_id: PeerId) {
+    /// If no addresses are reported for the peer by the behaviour, this is a noop.
+    ///
+    /// Returns `Some` with the connection ID of the first connection
+    /// attempt if a new dialing attempt has been initiated.
+    pub fn dial(me: &mut Self, peer_id: PeerId, condition: DialPeerCondition) -> Option<ConnectionId> {
```

Comment:
I'm not a fan of this API.
I agree that it is better than the current one, but in my opinion it would be better to have something more explicit.
I don't have any precise opinion, but a function that is called `dial` should *always* dial, without any condition attached to it.
If we want the user to be able to dial only under certain conditions, then we should provide an API similar to the one in `Network`, where you can first ask for the state of the connectivity with the peer.

Reply:
Ok, it sounds a bit like a stylistic preference which I'm not sure I can appreciate, but you are fine with the `DialPeerCondition` on `NetworkBehaviourAction::DialPeer`? Since the `NetworkBehaviour` has no means to figure out whether a peer is currently being dialed as it only interacts via callbacks, `Swarm::dial` is essentially modeled analogous to emitting `NetworkBehaviourAction::DialPeer`, thus showing both a certain symmetry on these APIs and reuse of data types, which to me seemed a good thing, but I'm not married to it.

Reply:
I reworked the `Swarm::dial` API with some related adjustments to make `Swarm::dial` and `Swarm::dial_addr` (as well as the underlying `Network` functions) more alike in https://github.com/libp2p/rust-libp2p/pull/1519/commits/e047ff93b488ae68a4191d1338fc63318654837b. There are some subtleties to preserve the existing behaviour of always adding the result of `addresses_of_peer` to the current dialing attempt, if there is one, though I don't think we can keep this behaviour for `Swarm::dial` if https://github.com/libp2p/rust-libp2p/pull/1506 is accepted without introducing new parameters to `Swarm::dial` due to ambiguity of meaning (if we want to distinguish "add addresses to current dialing attempt if there is one, otherwise start dialing" vs "always start new dialing attempt"; intuitively and for simplicity, https://github.com/libp2p/rust-libp2p/pull/1506 would have to always do the latter for `Swarm::dial`.)